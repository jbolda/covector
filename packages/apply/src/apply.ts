import { type Operation } from "effection";
import {
  writePkgFile,
  getPackageFileVersion,
  setPackageFileVersion,
  testSerializePkgFile,
} from "@covector/files";
import semver from "semver";

import type {
  PackageFile,
  CommonBumps,
  Releases,
  PackageCommand,
  DepTypes,
  Pkg,
  Logger,
} from "@covector/types";

export function* apply({
  logger,
  commands,
  allPackages,
  cwd = process.cwd(),
  bump = true,
  previewVersion = "",
  prereleaseIdentifier,
  logs = true,
}: {
  logger: Logger;
  commands: PackageCommand[];
  allPackages: Record<string, PackageFile>;
  cwd: string;
  bump: boolean;
  previewVersion: string;
  prereleaseIdentifier?: string;
  logs?: boolean;
}): Operation<PackageFile[]> {
  const changes = commands.reduce(
    (finalChanges: { [k: string]: PackageCommand }, command) => {
      finalChanges[command.pkg] = command;
      return finalChanges;
    },
    {}
  );

  const bumps = bumpAll({
    logger,
    changes,
    allPackages,
    previewVersion,
    logs,
    prereleaseIdentifier,
  });

  if (bump) {
    yield writeAll({
      bumps: bumps.reduce(
        (final: PackageFile[], current) =>
          !current.file ? final : final.concat([current]),
        []
      ),
      cwd,
    });
  } else {
    bumps.forEach((b) => {
      if (!!b && logs)
        logger.info(
          `${b.name} planned to be bumped from ${b.currentVersion} to ${b.version}`
        );
    });
  }
  return bumps;
}

export function* validateApply({
  logger,
  commands,
  allPackages,
  prereleaseIdentifier,
}: {
  logger: Logger;
  commands: PackageCommand[];
  allPackages: Record<string, PackageFile>;
  prereleaseIdentifier?: string;
}): Operation<true | Error> {
  const changes = commands.reduce(
    (finalChanges: { [k: string]: PackageCommand }, command) => {
      finalChanges[command.pkg] = command;
      return finalChanges;
    },
    {}
  );

  const bumps = bumpAll({
    logger,
    changes,
    allPackages,
    logs: false,
    prereleaseIdentifier,
  }).reduce(
    (final: PackageFile[], current) =>
      !current.file ? final : final.concat([current]),
    []
  );

  try {
    for (let bump of bumps) {
      testSerializePkgFile({ logger, packageFile: bump });
    }
    // will throw on validation error and not return true
    return true;
  } catch (e) {
    throw e;
  }
}

const writeAll = function* ({
  bumps,
  cwd,
}: {
  bumps: PackageFile[];
  cwd: string;
}) {
  for (let bump of bumps) {
    yield writePkgFile({ packageFile: bump, cwd });
  }
};

const bumpAll = ({
  logger,
  changes,
  allPackages,
  logs = true,
  previewVersion = "",
  prereleaseIdentifier,
}: {
  logger: Logger;
  changes: Releases;
  allPackages: Record<string, PackageFile>;
  logs?: boolean;
  previewVersion?: string;
  prereleaseIdentifier?: string;
}) => {
  // spread so that we can mutate
  let packageFiles = { ...allPackages };

  // loop through all packages and bump the main version for each
  for (let pkg of Object.keys(changes)) {
    if (!packageFiles[pkg]?.file || changes[pkg].type === "noop") continue;

    if (logs && !previewVersion) {
      logger.info(`bumping ${pkg} with ${changes[pkg].type}`);
    } else if (previewVersion) {
      // change log (assume that the prerelease will be removed)
      logger.info(
        `bumping ${pkg} with ${previewVersion} identifier to publish a preview`
      );
    }

    // bump the package's version number
    packageFiles[pkg] = bumpMain({
      packageFile: packageFiles[pkg],
      bumpType: changes[pkg].type,
      previewVersion,
      prereleaseIdentifier,
      errorOnVersionRange: changes[pkg].errorOnVersionRange,
    });
  }

  for (let pkg of Object.keys(changes)) {
    // bump any deps that are in the monorepo
    // and have a version bump as well
    if (changes?.[pkg]?.dependencies) {
      let deps = changes?.[pkg]?.dependencies || [];
      for (let pkgDep of deps) {
        if (!!changes[pkgDep]) {
          packageFiles[pkg] = bumpDeps({
            packageFile: packageFiles[pkg],
            dep: pkgDep,
            previewVersion,
            packageFiles,
          });
        }
      }
    }
  }

  return Object.keys(packageFiles)
    .filter((pkg) => changes?.[pkg])
    .map((pkg) => packageFiles[pkg]);
};

const bumpMain = ({
  packageFile,
  bumpType,
  previewVersion,
  prereleaseIdentifier,
  errorOnVersionRange,
}: {
  packageFile: PackageFile;
  bumpType: CommonBumps;
  previewVersion: string;
  prereleaseIdentifier?: string;
  errorOnVersionRange?: string;
}) => {
  let pkg = { ...packageFile };
  if (!pkg.version)
    throw new Error(`${pkg.name} does not have a version number.`);

  if (bumpType === "noop")
    throw new Error(`${pkg.name} needs a valid bump type, passed ${bumpType}`);

  if (prereleaseIdentifier && typeof prereleaseIdentifier !== "string")
    throw new Error(
      `${pkg.name} needs prereleaseIdentifier passed as a string`
    );

  let next = semver.inc(pkg.version, bumpType, prereleaseIdentifier);
  if (next) {
    pkg.version = next;
    pkg.versionMajor = semver.major(next);
    pkg.versionMinor = semver.minor(next);
    pkg.versionPatch = semver.patch(next);
    pkg.versionPrerelease = semver.prerelease(next);
  }
  const prevVersion = getPackageFileVersion({ pkg });
  const preVersionCleaned = semver.prerelease(prevVersion)
    ? semver.inc(prevVersion, "patch")
    : prevVersion;
  let version =
    previewVersion && previewVersion !== ""
      ? semver.valid(`${preVersionCleaned}-${previewVersion}`)
      : semver.inc(prevVersion, bumpType, prereleaseIdentifier);

  if (version) {
    pkg = setPackageFileVersion({ pkg, version });
    if (errorOnVersionRange && semver.satisfies(version, errorOnVersionRange)) {
      throw new Error(
        `${pkg.name} will be bumped to ${version}. ` +
          `This satisfies the range ${errorOnVersionRange} which the configuration disallows. ` +
          `Please adjust your bump to accommodate the range or otherwise adjust the allowed range in \`errorOnVersionRange\`.`
      );
    }
  }

  return pkg;
};

const bumpDeps = ({
  packageFile,
  dep,
  previewVersion,
  packageFiles,
}: {
  packageFile: PackageFile;
  dep: string;
  previewVersion: string;
  packageFiles: Record<string, PackageFile>;
}) => {
  let pkg = { ...packageFile };

  if (pkg.pkg && pkg.file) {
    const currentPkg = pkg.pkg;
    const depTypes: DepTypes[] = [
      "dependencies",
      "devDependencies",
      "dev-dependencies",
      "build-dependencies",
      "target",
    ];
    const depPkg = packageFiles[dep];
    const depName = depPkg.pkg.package?.name || depPkg.pkg.name || dep;
    depTypes.forEach((property: DepTypes) => {
      if (property && property in currentPkg) {
        if (property === "target") {
          const targets = currentPkg[property] as object;
          for (const target of Object.values(targets)) {
            depTypes.forEach((property: DepTypes) => {
              if (property && property in target) {
                const version = getDepBumpVersion({
                  pkg,
                  currentPkg: target,
                  property,
                  depName,
                  dep,
                  previewVersion,
                  packageFiles,
                  getPreviousVersion: () => target[property][depName]?.version,
                });
                if (version) {
                  target[property][depName].version = version;
                }
              }
            });
          }
        } else {
          const version = getDepBumpVersion({
            pkg,
            currentPkg,
            property,
            depName,
            dep,
            previewVersion,
            packageFiles,
            getPreviousVersion: () =>
              getPackageFileVersion({ pkg, property, dep: depName }),
          });
          if (version) {
            pkg = setPackageFileVersion({
              pkg,
              version,
              property,
              dep: depName,
            });
          }
        }
      }
    });
  }

  return pkg;
};

const getDepBumpVersion = ({
  pkg,
  currentPkg,
  property,
  depName,
  dep,
  previewVersion,
  packageFiles,
  getPreviousVersion,
}: {
  pkg: PackageFile;
  currentPkg: any;
  property: DepTypes;
  depName: string;
  dep: string;
  previewVersion: string;
  packageFiles: Record<string, PackageFile>;
  getPreviousVersion: () => string;
}) => {
  const pkgProperties = Object.keys(currentPkg[property] as object) as Array<
    keyof Pkg
  >;
  for (const existingDep of pkgProperties) {
    // if pkg is in dep list
    if (existingDep === depName) {
      const prevVersion = getPreviousVersion();
      const versionRequirementMatch = /[\^=~]/.exec(prevVersion);
      const versionRequirement = versionRequirementMatch
        ? versionRequirementMatch[0]
        : "";

      const version = deriveVersionConsideringPartials({
        dependency: dep,
        prevVersion,
        versionRequirement,
        previewVersion,
        packageFiles,
      });
      return version;
    }
  }
  return null;
};

const deriveVersionConsideringPartials = ({
  dependency,
  prevVersion,
  versionRequirement,
  previewVersion,
  packageFiles,
}: {
  dependency: string;
  prevVersion: string;
  versionRequirement: string;
  previewVersion: string;
  packageFiles: Record<string, PackageFile>;
}) => {
  if (previewVersion && previewVersion !== "") {
    const preVersionCleaned = semver.prerelease(prevVersion)
      ? semver.inc(prevVersion, "patch")
      : prevVersion;
    return semver.valid(`${preVersionCleaned}-${previewVersion}`);
  }

  const pkg = packageFiles[dependency];
  const { version, versionMajor, versionMinor, versionPatch } = pkg;

  if (!version) throw new Error(`${pkg.name} doesn't have a version?`);

  let depVersion = version;
  if (prevVersion.split(".").length === 2) {
    depVersion = `${versionMajor}.${versionMinor}`;
  } else if (prevVersion.split(".").length === 1) {
    depVersion = `${versionMajor}`;
  }

  return `${versionRequirement}${depVersion}`;
};
