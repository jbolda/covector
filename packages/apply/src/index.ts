import {
  readPkgFile,
  writePkgFile,
  getPackageFileVersion,
  setPackageFileVersion,
  testSerializePkgFile,
} from "@covector/files";
import semver from "semver";
import { cloneDeep } from "lodash";

import type {
  PackageFile,
  ConfigFile,
  CommonBumps,
  ChangeParsed,
  Releases,
  PackageCommand,
} from "@covector/types";

export const apply = function* ({
  commands,
  config,
  cwd = process.cwd(),
  bump = true,
  previewVersion = "",
  prereleaseIdentifier = null,
}: {
  commands: PackageCommand[];
  config: ConfigFile;
  cwd: string;
  bump: boolean;
  previewVersion: string;
  prereleaseIdentifier: string | null;
}): Generator<any, PackageFile[], any> {
  const changes = commands.reduce(
    (finalChanges: { [k: string]: PackageCommand }, command) => {
      finalChanges[command.pkg] = command;
      return finalChanges;
    },
    {}
  );

  // @ts-ignore since TS doesn't like yielding on a Promise
  let allPackages = yield readAll({ changes, config, cwd });
  const bumps = bumpAll({
    changes,
    allPackages,
    previewVersion,
    prereleaseIdentifier,
  });

  if (bump) {
    yield writeAll({
      bumps: bumps.reduce(
        (final: PackageFile[], current) =>
          !current.vfile ? final : final.concat([current]),
        []
      ),
    });
  } else {
    bumps.forEach((b) => {
      if (!!b) console.log(`${b.name} planned to be bumped to ${b.version}`);
    });
  }
  return bumps;
};

export const validateApply = async ({
  commands,
  config,
  cwd = process.cwd(),
  prereleaseIdentifier = null,
}: {
  commands: PackageCommand[];
  config: ConfigFile;
  cwd: string;
  prereleaseIdentifier: string | null;
}) => {
  const changes = commands.reduce(
    (finalChanges: { [k: string]: PackageCommand }, command) => {
      finalChanges[command.pkg] = command;
      return finalChanges;
    },
    {}
  );
  let allPackages = await readAll({ changes, config, cwd });

  const bumps = bumpAll({
    changes,
    allPackages,
    logs: false,
    prereleaseIdentifier,
  }).reduce(
    (final: PackageFile[], current) =>
      !current.vfile ? final : final.concat([current]),
    []
  );

  try {
    for (let bump of bumps) {
      testSerializePkgFile({ packageFile: bump });
    }
    // will throw on validation error and not return true
    return true;
  } catch (e) {
    throw e;
  }
};

const readAll = async ({
  changes,
  config,
  cwd = process.cwd(),
}: {
  changes: { [k: string]: { parents: string[] } };
  config: ConfigFile;
  cwd: string;
}): Promise<{ [k: string]: PackageFile }> => {
  let templateShell: PackageFile = {
    version: "",
    pkg: { name: "", version: "" },
  };
  let files = Object.keys(changes).reduce(
    (fileList: { [k: string]: PackageFile }, change) => {
      fileList[change] = { ...templateShell };
      if (changes[change].parents && changes[change].parents.length > 0)
        changes[change].parents.forEach(
          (parent) => (fileList[parent] = { ...templateShell })
        );
      return fileList;
    },
    {}
  );

  const pkgs: string[] = Object.keys(files);
  const pkgFiles = await Promise.all(
    Object.keys(files).map((pkg) =>
      !config.packages[pkg].path
        ? { name: pkg }
        : readPkgFile({
            cwd,
            pkgConfig: config.packages[pkg],
            nickname: pkg,
          })
    )
  );

  return pkgs.reduce((list, pkg, index) => {
    list[pkg] = pkgFiles[index];
    return list;
  }, files);
};

const writeAll = function* ({ bumps }: { bumps: PackageFile[] }) {
  for (let bump of bumps) {
    yield writePkgFile({ packageFile: bump });
  }
};

const resolveParents = ({ config }: { config: ConfigFile }) => {
  return Object.keys(config.packages).reduce(
    (parents: { [k: string]: string[] }, pkg) => {
      parents[pkg] = [];
      Object.keys(config.packages).forEach((parent) => {
        if (
          !!config.packages[parent].dependencies &&
          config.packages[parent].dependencies!.includes(pkg)
        )
          parents[pkg].push(parent);
      });
      return parents;
    },
    {}
  );
};

type Changed = {
  [k: string]: {
    parents: string[];
    type: CommonBumps;
    changes?: ChangeParsed[];
  };
};
export const changesConsideringParents = ({
  assembledChanges,
  config,
  prereleaseIdentifier = null,
}: {
  assembledChanges: {
    releases: Releases;
    changes: ChangeParsed[];
  };
  config: ConfigFile;
  prereleaseIdentifier: string | null;
}) => {
  const parents = resolveParents({ config });

  let changes = Object.keys(assembledChanges.releases).reduce(
    (list: Changed, change) => {
      list[change] = assembledChanges.releases[change];
      list[change].parents = parents[change];
      return list;
    },
    {}
  );

  return {
    releases: parentBump(changes, parents, prereleaseIdentifier),
    changes: assembledChanges.changes,
  };
};

const parentBump = (
  initialChanges: Changed,
  parents: any,
  prereleaseIdentifier: string | null
): Changed => {
  let changes = { ...initialChanges };
  let recurse = false;
  Object.keys(initialChanges).forEach((main) => {
    if (changes[main].parents.length > 0) {
      changes[main].parents.forEach((pkg) => {
        // pkg is the parent and main is the child
        if (!!changes[pkg]) {
          // if a change is planned on the parent
          // we don't need to plan a release
        } else {
          // if the parent doesn't have a release
          // add one to adopt the next version of it's child
          changes[pkg] = {
            ...cloneDeep(changes[main]),
            // prerelease will do bump the X in `-beta.X` if it is already a prerelease
            // or it will do a prepatch if it isn't a prerelease
            type: !prereleaseIdentifier ? "patch" : "prerelease",
          };
          if (changes[pkg].changes) {
            changes[pkg].changes!.forEach((parentChange) => {
              parentChange.meta.dependencies = `Bumped due to a bump in ${main}.`;
            });
          }
          changes[pkg].parents = parents[pkg];
          // we also need to presume recursion to update the parents' parents
          recurse = true;
        }
      });
    }
  });
  return recurse ? parentBump(changes, parents, prereleaseIdentifier) : changes;
};

const bumpAll = ({
  changes,
  allPackages,
  logs = true,
  previewVersion = "",
  prereleaseIdentifier = null,
}: {
  changes: Releases;
  allPackages: { [k: string]: PackageFile };
  logs?: boolean;
  previewVersion?: string;
  prereleaseIdentifier: string | null;
}) => {
  let packageFiles = { ...allPackages };
  for (let pkg of Object.keys(changes)) {
    if (!packageFiles[pkg].vfile || changes[pkg].type === "noop") continue;
    if (logs && !previewVersion)
      console.log(`bumping ${pkg} with ${changes[pkg].type}`);
    if (previewVersion)
      // change log (assume that the prerelease will be removed)
      console.log(
        `bumping ${pkg} with ${previewVersion} template to publish a preview`
      );
    packageFiles[pkg] = bumpMain({
      packageFile: packageFiles[pkg],
      bumpType: changes[pkg].type,
      previewVersion,
      prereleaseIdentifier,
      errorOnVersionRange: changes[pkg].errorOnVersionRange,
    });
    if (changes[pkg] && changes[pkg].dependencies) {
      let deps = changes[pkg].dependencies!;
      for (let pkgDep of deps) {
        if (!!changes[pkgDep]) {
          packageFiles[pkg] = bumpDeps({
            packageFile: packageFiles[pkg],
            dep: pkgDep,
            bumpType: changes[pkgDep].type,
            previewVersion,
            prereleaseIdentifier,
          });
        }
      }
    }
  }

  return Object.keys(packageFiles).map((pkg) => packageFiles[pkg]);
};

const bumpMain = ({
  packageFile,
  bumpType,
  previewVersion,
  prereleaseIdentifier = null,
  errorOnVersionRange,
}: {
  packageFile: PackageFile;
  bumpType: CommonBumps;
  previewVersion: string;
  prereleaseIdentifier: string | null;
  errorOnVersionRange?: string;
}) => {
  let pkg = { ...packageFile };
  if (!pkg.version)
    throw new Error(`${pkg.name} does not have a version number.`);
  // @ts-ignore TODO bumpType should be narrowed to meet ReleaseType
  let next = semver.inc(pkg.version, bumpType, prereleaseIdentifier);
  if (next) {
    pkg.version = next;
    pkg.versionMajor = semver.major(next);
    pkg.versionMinor = semver.minor(next);
    pkg.versionPatch = semver.patch(next);
    pkg.versionPrerelease = semver.prerelease(next);
  }
  const prevVersion = getPackageFileVersion({ pkg });
  const preVersionCleaned = semver.prerelease(prevVersion) ? semver.inc(prevVersion, 'patch') : prevVersion;
  let version =
    previewVersion && previewVersion !== ""
      ? semver.valid(`${preVersionCleaned}-${previewVersion}`)
      : // @ts-ignore TODO bumpType should be narrowed to meet ReleaseType
        semver.inc(prevVersion, bumpType, prereleaseIdentifier);
  if (version) {
    pkg = setPackageFileVersion({ pkg, version });
    if (errorOnVersionRange && semver.satisfies(version, errorOnVersionRange)) {
      throw new Error(
        `${pkg.name} will be bumped to ${version}. ` +
          `This satisfies the range ${errorOnVersionRange} which the configuration disallows. ` +
          `Please adjust your bump to accommodate the range or otherwise adjust the allowed range in \`errorOnVersionRange\`.`
      );
    }
  }
  return pkg;
};

const bumpDeps = ({
  packageFile,
  dep,
  bumpType,
  previewVersion,
  prereleaseIdentifier = null,
}: {
  packageFile: PackageFile;
  dep: string;
  bumpType: string;
  previewVersion: string;
  prereleaseIdentifier: string | null;
}) => {
  let pkg = { ...packageFile };

  if (pkg.pkg && pkg.vfile)
    ["dependencies", "devDependencies", "dev-dependencies"].forEach(
      (property) => {
        // @ts-ignore
        if (pkg.pkg[property]) {
          // @ts-ignore
          Object.keys(pkg.pkg[property]).forEach((existingDep) => {
            if (existingDep === dep) {
              const prevVersion = getPackageFileVersion({ pkg, property, dep });
              const preVersionCleaned = semver.prerelease(prevVersion) ? semver.inc(prevVersion, 'patch') : prevVersion;
              let version =
                previewVersion && previewVersion !== ""
                  ? semver.valid(`${preVersionCleaned}-${previewVersion}`)
                  : incConsideringPartials(
                      dep,
                      prevVersion,
                      // @ts-ignore TODO deal with ReleaseType
                      bumpType,
                      prereleaseIdentifier
                    );
              if (version)
                pkg = setPackageFileVersion({ pkg, version, property, dep });
            }
          });
        }
      }
    );

  return pkg;
};

const incConsideringPartials = (
  dependency: string,
  version: string,
  bumpType: string,
  prereleaseIdentifier: string | null
) => {
  if (semver.valid(version)) {
    // @ts-ignore TODO deal with ReleaseType
    return semver.inc(version, bumpType, prereleaseIdentifier);
  } else {
    if (prereleaseIdentifier !== null) {
      console.warn(
        `bump for ${dependency} skipped as ${version} is a range, and does not specifically include prereleases. ` +
          `Please pin to a major.minor.patch for a prerelease bump.`
      );
      return null;
    }
    try {
      const coerced = semver.coerce(version);
      if (!coerced)
        throw new Error(
          `Cannot bump ${version} with ${bumpType}. Is it a valid version number?`
        );
      // @ts-ignore TODO deal with ReleaseType
      const fullVersion = semver.inc(coerced, bumpType).split(".");
      if (version.split(".").length === 2) {
        return [fullVersion[0], fullVersion[1]].join(".");
      } else if (version.split(".").length === 1) {
        return fullVersion[0];
      } else {
        // failsafe is better than null
        return fullVersion.join(".");
      }
    } catch {
      // failsafe is better than null
      // @ts-ignore TODO what should we _really_ do here?
      return semver.inc(semver.coerce(version), bumpType);
    }
  }
};
