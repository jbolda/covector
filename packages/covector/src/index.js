const { covector } = require("./run");
module.exports.covector = covector;

const { cli } = require("./cli");
module.exports.cli = cli;
