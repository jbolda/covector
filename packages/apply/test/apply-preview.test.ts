import { apply } from "../src";
import { loadFile, readAllPkgFiles } from "@covector/files";
import { describe, it } from "../../../helpers/test-scope.ts";
import { expect } from "vitest";
import pino from "pino";
import * as pinoTest from "pino-test";
import fixtures from "fixturez";
const f = fixtures(__dirname);

const configDefaults = {
  changeFolder: ".changes",
};

describe("package file applies preview bump", () => {
  it("bumps single js json", function* () {
    const stream = pinoTest.sink();
    const logger = pino(stream);
    const jsonFolder = f.copy("pkg.js-single-json"); // 0.5.9

    const commands = [
      {
        dependencies: undefined,
        manager: "javascript",
        path: "./",
        pkg: "js-single-json-fixture",
        type: "minor",
        parents: {},
      },
    ];

    const config = {
      ...configDefaults,
      packages: {
        "js-single-json-fixture": {
          path: "./",
          manager: "javascript",
        },
      },
    };

    const allPackages = yield readAllPkgFiles({ config, cwd: jsonFolder });

    yield apply({
      logger,
      //@ts-expect-error
      commands,
      config,
      cwd: jsonFolder,
      allPackages,
      previewVersion: "branch-name.12345",
    });
    const modifiedFile = yield loadFile("package.json", jsonFolder);
    expect(modifiedFile.content).toBe(
      "{\n" +
        '  "private": true,\n' +
        '  "name": "js-single-json-fixture",\n' +
        '  "description": "A single package at the root. No monorepo setup.",\n' +
        '  "repository": "https://www.github.com/jbolda/covector.git",\n' +
        '  "version": "0.5.9-branch-name.12345"\n' +
        "}\n"
    );

    yield pinoTest.consecutive(stream, [
      {
        msg: "bumping js-single-json-fixture with branch-name.12345 identifier to publish a preview",
        level: 30,
      },
    ]);
  });

  it("bumps multi js json", function* () {
    const stream = pinoTest.sink();
    const logger = pino(stream);
    const jsonFolder = f.copy("pkg.js-yarn-workspace"); // 1.0.0

    const commands = [
      {
        dependencies: ["yarn-workspace-base-pkg-b", "all"],
        manager: "javascript",
        path: "./",
        pkg: "yarn-workspace-base-pkg-a",
        type: "minor",
        parents: {},
      },
      {
        dependencies: undefined,
        manager: "javascript",
        path: undefined,
        pkg: "yarn-workspace-base-pkg-b",
        type: "minor",
        parents: { "yarn-workspace-base-pkg-a": "null" },
      },
      {
        dependencies: undefined,
        manager: "javascript",
        path: undefined,
        pkg: "all",
        type: "minor",
        parents: {
          "yarn-workspace-base-pkg-a": "null",
          "yarn-workspace-base-pkg-b": "null",
        },
      },
    ];

    const config = {
      ...configDefaults,
      packages: {
        "yarn-workspace-base-pkg-a": {
          path: "./packages/pkg-a/",
          manager: "javascript",
          dependencies: ["yarn-workspace-base-pkg-b", "all"],
        },
        "yarn-workspace-base-pkg-b": {
          path: "./packages/pkg-b/",
          manager: "javascript",
          dependencies: ["all"],
        },
        all: { version: true },
      },
    };

    const allPackages = yield readAllPkgFiles({ config, cwd: jsonFolder });

    yield apply({
      logger,
      //@ts-expect-error
      commands,
      config,
      allPackages,
      cwd: jsonFolder,
      previewVersion: "branch-name.12345",
    });
    const modifiedPkgAFile = yield loadFile(
      "packages/pkg-a/package.json",
      jsonFolder
    );
    expect(modifiedPkgAFile.content).toBe(
      "{\n" +
        '  "name": "yarn-workspace-base-pkg-a",\n' +
        '  "version": "1.0.0-branch-name.12345",\n' +
        '  "dependencies": {\n' +
        '    "yarn-workspace-base-pkg-b": "1.0.0-branch-name.12345"\n' +
        "  }\n" +
        "}\n"
    );

    const modifiedPkgBFile = yield loadFile(
      "packages/pkg-b/package.json",
      jsonFolder
    );
    expect(modifiedPkgBFile.content).toBe(
      "{\n" +
        '  "name": "yarn-workspace-base-pkg-b",\n' +
        '  "version": "1.0.0-branch-name.12345"\n' +
        "}\n"
    );

    yield pinoTest.consecutive(stream, [
      {
        msg: "bumping yarn-workspace-base-pkg-a with branch-name.12345 identifier to publish a preview",
        level: 30,
      },
      {
        msg: "bumping yarn-workspace-base-pkg-b with branch-name.12345 identifier to publish a preview",
        level: 30,
      },
      {
        msg: "bumping all with branch-name.12345 identifier to publish a preview",
        level: 30,
      },
    ]);
  });
});

describe("package file applies preview bump to pre-release", () => {
  it("bumps single js json without pre-release", function* () {
    const stream = pinoTest.sink();
    const logger = pino(stream);
    const jsonFolder = f.copy("pkg.js-single-prerelease-json"); // 0.5.9-abc.2

    const commands = [
      {
        dependencies: undefined,
        manager: "javascript",
        path: "./",
        pkg: "js-single-prerelease-json-fixture",
        type: "minor",
        parents: {},
      },
    ];

    const config = {
      ...configDefaults,
      packages: {
        "js-single-prerelease-json-fixture": {
          path: "./",
          manager: "javascript",
        },
      },
    };

    const allPackages = yield readAllPkgFiles({ config, cwd: jsonFolder });

    yield apply({
      logger,
      //@ts-expect-error
      commands,
      config,
      allPackages,
      cwd: jsonFolder,
      previewVersion: "branch-name.12345",
    });
    const modifiedFile = yield loadFile("package.json", jsonFolder);
    expect(modifiedFile.content).toBe(
      "{\n" +
        '  "private": true,\n' +
        '  "name": "js-single-prerelease-json-fixture",\n' +
        '  "description": "A single package at the root. No monorepo setup.",\n' +
        '  "version": "0.5.9-branch-name.12345"\n' +
        "}\n"
    );

    yield pinoTest.consecutive(stream, [
      {
        msg: "bumping js-single-prerelease-json-fixture with branch-name.12345 identifier to publish a preview",
        level: 30,
      },
    ]);
  });

  it("bumps multi js json without pre-release", function* () {
    const stream = pinoTest.sink();
    const logger = pino(stream);
    const jsonFolder = f.copy("pkg.js-yarn-prerelease-workspace");

    const commands = [
      {
        dependencies: ["yarn-workspace-base-pkg-b", "all"],
        manager: "javascript",
        path: "./",
        pkg: "yarn-workspace-base-pkg-a",
        type: "minor",
        parents: {},
      },
      {
        dependencies: undefined,
        manager: "javascript",
        path: undefined,
        pkg: "yarn-workspace-base-pkg-b",
        type: "minor",
        parents: { "yarn-workspace-base-pkg-a": "null" },
      },
      {
        dependencies: undefined,
        manager: "javascript",
        path: undefined,
        pkg: "all",
        type: "minor",
        parents: {
          "yarn-workspace-base-pkg-a": "null",
          "yarn-workspace-base-pkg-b": "null",
        },
      },
    ];

    const config = {
      packages: {
        "yarn-workspace-base-pkg-a": {
          path: "./packages/pkg-a/", // 1.0.0-abc.2
          manager: "javascript",
          dependencies: ["yarn-workspace-base-pkg-b", "all"],
        },
        "yarn-workspace-base-pkg-b": {
          path: "./packages/pkg-b/", // 1.0.0-abc.3
          manager: "javascript",
          dependencies: ["all"],
        },
        all: { version: true },
      },
    };

    const allPackages = yield readAllPkgFiles({ config, cwd: jsonFolder });

    yield apply({
      logger,
      //@ts-expect-error
      commands,
      config,
      allPackages,
      cwd: jsonFolder,
      previewVersion: "branch-name.12345",
    });
    const modifiedPkgAFile = yield loadFile(
      "packages/pkg-a/package.json",
      jsonFolder
    );
    expect(modifiedPkgAFile.content).toBe(
      "{\n" +
        '  "name": "yarn-workspace-base-pkg-a",\n' +
        '  "version": "1.0.0-branch-name.12345",\n' +
        '  "dependencies": {\n' +
        '    "yarn-workspace-base-pkg-b": "1.0.0-branch-name.12345"\n' +
        "  }\n" +
        "}\n"
    );

    const modifiedPkgBFile = yield loadFile(
      "packages/pkg-b/package.json",
      jsonFolder
    );
    expect(modifiedPkgBFile.content).toBe(
      "{\n" +
        '  "name": "yarn-workspace-base-pkg-b",\n' +
        '  "version": "1.0.0-branch-name.12345"\n' +
        "}\n"
    );

    yield pinoTest.consecutive(stream, [
      {
        msg: "bumping yarn-workspace-base-pkg-a with branch-name.12345 identifier to publish a preview",
        level: 30,
      },
      {
        msg: "bumping yarn-workspace-base-pkg-b with branch-name.12345 identifier to publish a preview",
        level: 30,
      },
      {
        msg: "bumping all with branch-name.12345 identifier to publish a preview",
        level: 30,
      },
    ]);
  });
});
