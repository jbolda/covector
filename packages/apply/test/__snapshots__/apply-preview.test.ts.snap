// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`package file applies preview bump bumps multi js json 1`] = `
Object {
  "consoleDir": Array [],
  "consoleLog": Array [
    Array [
      "bumping yarn-workspace-base-pkg-a with branch-name.12345 identifier to publish a preview",
    ],
    Array [
      "bumping yarn-workspace-base-pkg-b with branch-name.12345 identifier to publish a preview",
    ],
    Array [
      "bumping all with branch-name.12345 identifier to publish a preview",
    ],
  ],
}
`;

exports[`package file applies preview bump bumps single js json 1`] = `
Object {
  "consoleDir": Array [],
  "consoleLog": Array [
    Array [
      "bumping js-single-json-fixture with branch-name.12345 identifier to publish a preview",
    ],
  ],
}
`;

exports[`package file applies preview bump to pre-release bumps multi js json without pre-release 1`] = `
Object {
  "consoleDir": Array [],
  "consoleLog": Array [
    Array [
      "bumping yarn-workspace-base-pkg-a with branch-name.12345 identifier to publish a preview",
    ],
    Array [
      "bumping yarn-workspace-base-pkg-b with branch-name.12345 identifier to publish a preview",
    ],
    Array [
      "bumping all with branch-name.12345 identifier to publish a preview",
    ],
  ],
}
`;

exports[`package file applies preview bump to pre-release bumps single js json without pre-release 1`] = `
Object {
  "consoleDir": Array [],
  "consoleLog": Array [
    Array [
      "bumping js-single-prerelease-json-fixture with branch-name.12345 identifier to publish a preview",
    ],
  ],
}
`;
